labels_test.go

```diff
@@ -0,0 +1,149 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import (
+	"regexp"
+	"sync"
+	"testing"
+	"testing/quick"
+
+	"github.com/prometheus/common/model"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestDigester(t *testing.T) {
+	const (
+		goroutines = 10
+		iterations = 100
+	)
+
+	expected := []byte{'f', 'o', 'o', 0, 'b', 'a', 'r'}
+
+	var wg sync.WaitGroup
+	for i := 0; i < goroutines; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for i := 0; i < iterations; i++ {
+				d := newDigester()
+				defer d.free()
+
+				assert.Equal(t, 0, len(d.digest()), "Expected fresh digester to have no internal state.")
+				assert.True(t, cap(d.digest()) > 0, "Expected fresh digester to have available capacity.")
+
+				d.add("", "foo")
+				d.add("", "bar")
+				assert.Equal(
+					t,
+					string(expected),
+					string(d.digest()),
+					"Expected digest to be null-separated concatenation of inputs.",
+				)
+			}
+		}()
+	}
+
+	wg.Wait()
+}
+
+func TestIsValidName(t *testing.T) {
+	// Regexp is a slower, but more easily verifiable, description of the Tally
+	// name specification.
+	tallyRe := regexp.MustCompile(`^[0-9A-z_\-]+$`)
+	isValid := func(s string) bool {
+		prom := model.IsValidMetricName(model.LabelValue(s)) && model.LabelName(s).IsValid()
+		tally := tallyRe.MatchString(s)
+		return prom && tally
+	}
+
+	assert.NoError(t, quick.CheckEqual(
+		isValid,
+		IsValidName,
+		nil, /* config */
+	), "Hand-rolled validation doesn't match Tally regexp && stock Prometheus validators.")
+}
+
+func TestIsValidLabelValue(t *testing.T) {
+	tallyRe := regexp.MustCompile(`^[0-9A-z_.\-]+$`)
+	isValid := func(s string) bool {
+		prom := model.LabelValue(s).IsValid()
+		tally := tallyRe.MatchString(s)
+		return prom && tally
+	}
+
+	assert.NoError(t, quick.CheckEqual(
+		isValid,
+		IsValidName,
+		nil, /* config */
+	), "Hand-rolled validation doesn't match Tally regexp && stock Prometheus validators.")
+}
+
+func TestScrubLabelValue(t *testing.T) {
+	tests := []struct {
+		in  string
+		out string
+	}{
+		{"foo", "foo"},
+		{"", DefaultLabelValue},
+		{"foo!", "foo_"},
+		{"!foo", "_foo"},
+		{"fOo1.!FoO", "fOo1._FoO"},
+	}
+
+	for _, tt := range tests {
+		assert.Equal(t, tt.out, scrubLabelValue(tt.in), "Unexpected result from scrubLabelValue.")
+	}
+}
+
+func TestScrubName(t *testing.T) {
+	tests := []struct {
+		in  string
+		out string
+	}{
+		{"", DefaultLabelValue},
+		{"foo", "foo"},
+		{"foo!", "foo_"},
+		{"!foo", "_foo"},
+		{"fOo1_.!FoO", "fOo1___FoO"},
+		{"FOo1.", "FOo1_"},
+		{"_Oo1.", "_Oo1_"},
+	}
+
+	for _, tt := range tests {
+		assert.Equal(t, tt.out, scrubName(tt.in), "Unexpected result from scrubName.")
+	}
+}
+
+func TestScrubQuick(t *testing.T) {
+	// Generative tests to ensure that scrubbing always produces valid names and
+	// values.
+	t.Run("scrubName", func(t *testing.T) {
+		quick.Check(func(s string) bool {
+			return IsValidName(scrubName(s))
```

Comment:
we should ensure that `scrubName` should only modifies names if they're not valid to begin with. This will ensure that `scrubNames` is not being over-aggressive

e.g.,:

```go
if IsValidName(s) {
  return s == scrubName(s))
} else {
  return IsValidName(scrubName(s))
}
```

Reply:
Good point, done.